examples/frontend-links/src/frontend/core.cljs

```diff
@@ -0,0 +1,143 @@
+(ns frontend.core
+  (:require [clojure.string :as string]
+            [fipp.edn :as fedn]
+            [reagent.core :as r]
+            [reitit.coercion :as rc]
+            [reitit.coercion.spec :as rss]
+            [reitit.frontend :as rf]
+            [reitit.frontend.easy :as rfe]
+            [spec-tools.data-spec :as ds]))
+
+;; Components similar to react-router `Link`, `NavLink` and `Redirect`
+;; with Reitit frontend.
+
+(defn home-page []
+  [:div
+   [:h2 "Welcome to frontend"]
+   [:p "This is home page"]])
+
+(defn about-page []
+  [:div
+   [:h2 "About frontend"]
+   [:p "This is about page"]])
+
+(defn redirect!
+  "If `push` is truthy, previous page will be left in history."
+  [{:keys [to path-params query-params push]}]
+  (if push
+    (rfe/push-state to path-params query-params)
+    (rfe/replace-state to path-params query-params)))
+
+(defn Redirect
+  "Component that only causes a redirect side-effect."
+  [props]
+  (redirect! props)
+  nil)
+
+(defn item-page [match]
+  (let [{:keys [path query]} (:parameters match)
+        {:keys [id]} path]
+    (if (< id 1)
+      [Redirect {:to ::frontpage}]
+      [:div
+       [:h2 "Selected item " id]
+       (when (:foo query)
+         [:p "Optional foo query param: " (:foo query)])])))
+
+(def routes
+  [["/"
+    {:name ::frontpage
+     :view home-page}]
+
+   ["/about"
+    {:name ::about
+     :view about-page}]
+
+   ["/item/:id"
+    {:name ::item
+     :view item-page
+     :parameters
+     {:path {:id int?}
+      :query {(ds/opt :foo) keyword?}}}]])
+
+(def router
+  (rf/router routes {:data {:coercion rss/coercion}}))
+
+(defonce current-match (r/atom nil))
+
+(defn- resolve-href
+  [to path-params query-params]
+  (if (keyword? to)
+    (rfe/href to path-params query-params)
+    (let [match  (rf/match-by-path router to)
+          route  (-> match :data :name)
+          params (not-empty (or path-params (:path-params match)))
```

Comment:
These `not-empty` calls are probably unnecessary?

Reply:
I added them because empty query params `{}` cause redundant `?` in href. Maybe it could be fixed in `rfe/href`?

Reply:
The problems seems to be in `reitit.core/match->path`: https://github.com/metosin/reitit/blob/master/modules/reitit-core/src/reitit/core.cljc#L70

You could fix it in this PR, or open new issue for this problem & remove the workaround in the example.

Reply:
I decided to fix it and added the test case. I weighted wether I'd use `not-empty` or `seq`. Both worked but I decided to use `not-empty` in case someone in the future still expects to get a map instead of vectors downstream.